planner/cascades/testdata/transformation_rules_suite_out.json

```diff
@@ -229,5 +229,30 @@
         ]
       }
     ]
+  },
+  {
+    "Name": "TestProjectionElimination",
+    "Cases": [
+      {
+        "SQL": "select a from (select a, b from (select a, b, c from t) as t2) as t3",
+        "Result": [
+          "Group#0 Schema:[test.t.a], UniqueKey:[test.t.a]",
+          "    Projection_4 input:[Group#1], test.t.a",
+          "Group#1 Schema:[test.t.a], UniqueKey:[test.t.a]",
+          "    TableScan_1 table:t, pk col:test.t.a"
+        ]
+      },
+      {
+        "SQL": "select a from (select floor(a) as a, b, c from t) as t2",
+        "Result": [
+          "Group#0 Schema:[Column#13]",
+          "    Projection_3 input:[Group#1], Column#13",
+          "Group#1 Schema:[Column#13]",
```

Comment:
This rule should merge this two projection in fact.

Reply:
I used to think so. But the following [block](https://github.com/pingyu/tidb/blob/6a2bb1ca706ffcc6efa6fcca9bb6d726562b16d8/planner/cascades/transformation_rules.go#L622) prevents this, and the codes are migrated from [`canProjectionBeEliminatedLoose`](https://github.com/pingcap/tidb/blob/4b5ae3a34c328742354483bb4fe3dd0cdd443888/planner/core/rule_eliminate_projection.go#L24) of `rule_eliminate_project.go`

```go
	replace := make(map[string]*expression.Column)
	for i, col := range child.Schema().Columns {
		if col1, ok := child.Exprs[i].(*expression.Column); ok {
			replace[string(col.HashCode(nil))] = col1
		} else {
			return nil, false, false, nil
		}
	}
```

So it is safe to eliminate the projection with expressions more than single columns ?

Reply:
This rule you added is to merge two projection, instead of eliminating one of them. It's safe to do this in this rule.

Reply:
OK, I see.